File: be/src/domain/use-cases/UserUseCases.ts
import IUserRepository from "@src/domain/interfaces/IUserRepository";
import { User } from "@domain/entities";
const jwt = require('jsonwebtoken');
const bcrypt = require('bcrypt');
require('dotenv').config();
export default class UserUseCases {
    constructor(
        private userRepository: IUserRepository,
    ) {}
    async getUserByEmail(email: string): Promise<User | null> {
        const result = await this.userRepository.getUserByEmail(email);
        if (result) {
            return result.user;
        }
        return null;
    }
    async getAllUsers(): Promise<User[]> {
        return this.userRepository.getAllUsers();
    }
    async getUserById(id: string): Promise<User | null> {
        return this.userRepository.getUserById(id);
    }
    async updateUser(user: {id:string, email?:string, username?:string }): Promise<User | null> {
        return this.userRepository.updateUser(user);
    }
    async updateUserPassword(id: string, password: string): Promise<User | null> {
        const hashedPassword = await bcrypt.hash(password, process.env.SALT||10);
        return this.userRepository.updatePassword(id, hashedPassword);
    }
    async deleteUser(id: string): Promise<boolean> {
        return this.userRepository.deleteUser(id);
    }
}

File: be/internal/domain/usecases/UserUseCases.go
package usecases

import (
	"root/internal/domain/entities"
	"root/internal/domain/interfaces"

	"golang.org/x/crypto/bcrypt"
)

type UserUseCases struct {
	UserRepo interfaces.UserRepository
}

func NewUserUseCases(userRepo interfaces.UserRepository) *UserUseCases {
	return &UserUseCases{UserRepo: userRepo}
}

func (usecases *UserUseCases) GetUserByEmail(email string) (*entities.User, error) {
	user, _, err := usecases.UserRepo.GetUserByEmail(email)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func (usecases *UserUseCases) GetAllUsers() ([]entities.User, error) {
	users, err := usecases.UserRepo.GetAllUsers()
	if err != nil {
		return nil, err
	}
	return users, nil
}

func (usecases *UserUseCases) GetUserById(id string) (*entities.User, error) {
	user, err := usecases.UserRepo.GetUserById(id)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func (usecases *UserUseCases) UpdateUser(id string, changes *entities.UserChanges) (*entities.User, error) {

	user, err := usecases.UserRepo.UpdateUser(id, changes)
	if err != nil {
		return nil, err
	}
	return user, nil
}

func (usecases *UserUseCases) UpdatePassword(id string, password string) (*entities.User, error) {
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}
	user, err := usecases.UserRepo.UpdatePassword(id, string(hashedPassword))
	if err != nil {
		return nil, err
	}
	return user, nil
}

func (usecases *UserUseCases) DeleteUser(id string) error {
	return usecases.UserRepo.DeleteUser(id)
}

func (usecases *UserUseCases) GetUserByName(name string) ([]entities.User, error) {
	users, err := usecases.UserRepo.GetUserByName(name)
	if err != nil {
		return nil, err
	}
	return users, nil
}
